diff -rNu original-0.11/www/inc/www.mattb.inc.php /home/httpd/www.mattb.net.nz/html/original/inc/www.mattb.inc.php
--- original-0.11/www/inc/www.mattb.inc.php	1970-01-01 12:00:00.000000000 +1200
+++ /home/httpd/www.mattb.net.nz/html/original/inc/www.mattb.inc.php	2005-11-17 01:15:44.000000000 +1300
@@ -0,0 +1,27 @@
+<?php
+require("/home/httpd/www.mattb.net.nz/html/inc/header.php");
+
+class mattb_www extends C_www {
+    
+    function header($title) {
+        
+        parent::header($title);
+
+        /* mattb.net.nz specific page layout */
+        display_header();
+        echo "<div id=\"sidebar\">\n";
+        display_menu("gallery");
+        echo "</div>\n";
+        display_copyright();
+
+        echo "<div id=\"body\">\n";
+    }
+ 
+    function footer() {
+        echo "</div>\n";
+        echo "</body>\n";
+        echo "</html>\n";
+    }
+   
+}
+?>
diff -rNu original-0.11/www/index.php /home/httpd/www.mattb.net.nz/html/original/index.php
--- original-0.11/www/index.php	2005-11-14 13:53:21.000000000 +1300
+++ /home/httpd/www.mattb.net.nz/html/original/index.php	2005-11-17 01:16:38.000000000 +1300
@@ -14,6 +14,7 @@
 require_once ("lib/lib.l10n.php");
 require_once("inc/config.inc.php");
 require_once("inc/www.class.inc.php");
+require_once("inc/www.mattb.inc.php");
 require_once("inc/funkce.inc.php");
 //session_name("navstevnik");
 //session_register("page");
@@ -52,7 +53,7 @@
 	$exif_style="descriptive";
 } */
 
-$page = new C_www;
+$page = new mattb_www;
 //default colors
 
 if (!is_dir("$gallery_dir/$galerie/thumbs")) {
