diff -u cfengine2-2.1.21/debian/changelog cfengine2-2.1.21/debian/changelog
--- cfengine2-2.1.21/debian/changelog
+++ cfengine2-2.1.21/debian/changelog
@@ -1,3 +1,10 @@
+cfengine2 (2.1.21-1crcnet1) crcnet; urgency=low
+
+  * Rebuilt for squeeze.
+  * Build-deps tweaked to use libdb4.6.
+
+ -- Matt Brown <mattb@mattb.net.nz>  Thu, 08 Mar 2012 20:08:54 +0000
+
 cfengine2 (2.1.21-1crcnet0) crcnet; urgency=low
 
   * New upstream release
diff -u cfengine2-2.1.21/debian/control cfengine2-2.1.21/debian/control
--- cfengine2-2.1.21/debian/control
+++ cfengine2-2.1.21/debian/control
@@ -2,7 +2,7 @@
 Section: admin
 Priority: optional
 Maintainer: Matt Brown <matt@crc.net.nz>
-Build-Depends: bison, debhelper (>= 4.1.76), flex, libdb4.2-dev, openssl, libssl-dev, perl, dbs, autotools-dev, autoconf, automake1.9, tetex-bin, texinfo
+Build-Depends: bison, debhelper (>= 4.1.76), flex, libdb4.6-dev, openssl, libssl-dev, perl, dbs, autotools-dev, autoconf, automake1.9, texinfo
 Build-Conflicts: automake1.4
 Standards-Version: 3.6.1
 
